odoo/addons/base/models/ir_actions_report.py

```diff
@@ -772,13 +773,6 @@ def _render_qweb_pdf(self, res_ids=None, data=None):
         # https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2083
         context['debug'] = False
 
-        # The test cursor prevents the use of another environment while the current
-        # transaction is not finished, leading to a deadlock when the report requests
-        # an asset bundle during the execution of test scenarios. In this case, return
-        # the html version.
```

Comment:
Has this condition disappeared? Or is it assumed to be redundant with the test_enable conf of line 752?

@rco-odoo if we're going to assume test_mode ⊇ TestCursor then shouldn't we assert that in `enter_test_mode`, and document it?

Reply:
It is assumed to be redundant with the test_enable conf of line 752

Reply:
The purpose of the "test mode" on the registry is to make external calls (HTML, RPC) proceed on the database but in such a way that (1) requests behave in a transactional way, (2) a request can "see" the changes made by former requests, and (3) all those requests can be rolled back.  The test mode was introduced for running tours as regular tests.

When in test mode, `registry.cursor()` returns a `TestCursor` instead of a regular `Cursor`.  The `TestCursor` implements a sub-transaction inside a regular `Cursor`, so that all the sub-transactions made while in test mode are actually serialized inside one actual transaction, which may be rolled back (because tests don't commit).  Test cursors are not aimed at being used outside the registry test mode, and the latter is aimed at being used in an `HttpCase`.

You can consider that: `TestCursor` ⊂ `test_mode` ⊂ `config['test_enable'] or config['test_file']`.  The implications are strict, since the code may be called directly from a test method with a regular cursor.  Also, the flag `test_enable` may be set while running code outside a test (like code being run while installing a module).

Reply:
In practice `test_mode` ≈ `HttpCase` (they correspond in most cases).

Reply:
Note that it wasn't just redundant but also wrong, going through that code would raise a TB about tuple not being returned as described in commit message.
That's mainly why I removed it.